flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/async/Emitter.java

```diff
@@ -80,9 +88,14 @@ public void run() {
 		try {
 			while (running) {
 				LOG.debug("Wait for next completed async stream element result.");
-				AsyncResult streamElementEntry = streamElementQueue.peekBlockingly();
-
-				output(streamElementEntry);
+				AsyncResult asyncResult = streamElementQueue.peekBlockingly();
+				executor.submit(() -> {
+					try {
+						output(asyncResult);
+					} catch (InterruptedException e) {
```

Comment:
Why do we need this `InterruptedException`? I see that it's coming (and it is the same in master branch) from `streamElementQueue.poll()`, however I don't understand it:
1. We shouldn't be blocking neither in the mailbox thread (this version) nor under the checkpoint lock (master version), so this already looks suspicious.
2. `poll` method that blocks is already "mildly" surprising contract on it's own
3. From the code it looks like this `poll` call should never block and changing it to a proper poll (returning `Optional` or `null`) and adding a check state that the element is present, would work just as well but would be better/cleaner.

@AHeise I remember that you were also rising some objections about the `Emitter` code and that it could be simplified and you wanted to work on that later. Did you have this in mind as well?

Am I missing something? If not can you @AHeise create a JIRA ticket for cleaning up those things?

Reply:
Yes, I would like to get rid of any blocking methods inside emitter and `streamElementQueue`.
The main idea is to transfer the results of async operations via letters to the mailbox thread and then only have non-blocking, non-synchronized data structures that ensure the ordering of inputs/results. At that point, no interruptions can occur.
I'm opening a ticket for that.

Reply:
https://issues.apache.org/jira/browse/FLINK-14044